### PR TITLE
more #define updates

### PR DESCRIPTION
**Pull request by Goober5000**
Convert `REF_TYPE_*`, `SEXP_MODE_*`, and `SEXP_SRC_*` from bitfield flags to enum classes.

### CODE REVIEW

File: code/parse/sexp.h [hunk 1]
```diff
@@ -979,28 +979,36 @@ const char *CTEXT(int n);
...
+	WING_ORDER,
+	DEBRIEFING,
+	BRIEFING,
+	UNKNOWN
+};
+
+enum class SEXP_MODE
+{
+	GENERAL = 0,
+	CAMPAIGN
```
Comment (BMagnu) [inline code/parse/sexp.h:1010]: Almost all scoped enums in FSO, at least as far as I am aware, have a snake_case class name and SCREAMING_SNAKE_CASE value names, so I'd prefer if the enum class names here were made lowercase

File: fred2/management.cpp [hunk 2]
```diff
@@ -1911,77 +1911,80 @@ int advanced_stricmp(char *one, char *two)
 // returns 0: go ahead change object
 //			  1: don't change it
 //			  2: abort (they used cancel to go to reference)
-int reference_handler(char *name, int type, int obj)
+int reference_handler(char *name, SEXP_REF_TYPE type, int obj)
 {
 	char msg[2048], text[128], type_name[128];
 	int r, n, node;
+	SEXP_SRC source;
```
Comment (BMagnu) [inline fred2/management.cpp:1918]: Why declare source up here?
Declarations at the function start are obsolete anyways, so I'd suggest moving it and the declaration for n down to when they're assigned.

File: fred2/management.cpp [hunk 3]
```diff
@@ -2082,9 +2087,9 @@ int orders_reference_handler(int code, char *msg)
 
 	ShipGoalsDlg dlg_goals;
 
-	n = code & SRC_DATA_MASK;
-	switch (code & SRC_MASK) {
-		case SRC_SHIP_ORDER:
+	n = source_index;
```
Comment (BMagnu) [inline fred2/management.cpp:2090]: Well, this is now obsolete. n is just source index, so just using source_index directly is cleaner here

File: fred2/management.cpp [hunk 4]
```diff
@@ -2126,9 +2131,10 @@ int sexp_reference_handler(int node, int code, char *msg)
 	if (r == IDYES)
 		return 0;
 
-	switch (code & SRC_MASK) {
-		case SRC_SHIP_ARRIVAL:
-		case SRC_SHIP_DEPARTURE:
+	n = source_index;
```
Comment (BMagnu) [inline fred2/management.cpp:2134]: same here

File: qtfred/src/mission/Editor.cpp [hunk 5]
```diff
@@ -995,57 +995,60 @@ void Editor::setActiveViewport(EditorViewport* viewport) {
 	_lastActiveViewport = viewport;
 }
 
-int Editor::reference_handler(const char* name, int type, int obj) {
+int Editor::reference_handler(const char* name, SEXP_REF_TYPE type, int obj) {
 
 	char msg[2048], text[128], type_name[128];
 	int r, n, node;
+	SEXP_SRC source;
```
Comment (BMagnu) [inline qtfred/src/mission/Editor.cpp:1002]: See above

File: qtfred/src/mission/Editor.cpp [hunk 6]
```diff
@@ -1181,9 +1186,9 @@ int Editor::orders_reference_handler(int  /*code*/, char* msg) {
 	/*
 	ShipGoalsDlg dlg_goals;
 
-	auto n = code & SRC_DATA_MASK;
-	switch (code & SRC_MASK) {
-	case SRC_SHIP_ORDER:
+	auto n = source_index;
```
Comment (BMagnu) [inline qtfred/src/mission/Editor.cpp:1189]: see above

File: qtfred/src/mission/Editor.cpp [hunk 7]
```diff
@@ -1231,9 +1236,10 @@ int Editor::sexp_reference_handler(int  /*node*/, int  /*code*/, const char* msg
 
 	// TODO: add a generic dialog system for showing these dialogs
 	/*
-	switch (code & SRC_MASK) {
-	case SRC_SHIP_ARRIVAL:
-	case SRC_SHIP_DEPARTURE:
+	int n = source_index;
```
Comment (BMagnu) [inline qtfred/src/mission/Editor.cpp:1239]: same here

File: code/ai/aigoals.cpp [hunk 8]
```diff
@@ -2547,7 +2547,7 @@ void ai_update_goal_references(ai_goal *goals, int type, const char *old_name, c
 				break;
 
 			default:
-				Warning(LOCATION, "unhandled FRED reference type %d in ai_update_goal_references", type);
+				Warning(LOCATION, "unhandled FRED reference type %d in ai_update_goal_references", (int)type);
```
Comment (BMagnu) [inline code/ai/aigoals.cpp:2550]: _scoped_ enum casts should really be done using static_cast only

**Review by BMagnu — CHANGES_REQUESTED** [review]
Some small comments

File: code/parse/sexp.h (hunk 1 above)
Comment (Goober5000) [reply code/parse/sexp.h:1010]: As discussed, the name cases are a bit more mixed, but the ones in sexp.h at least are snake_case.  I've made the change.

File: fred2/management.cpp (hunk 2 above)
Comment (Goober5000) [reply fred2/management.cpp:1918]: To be consistent with the other variables, but changed

File: fred2/management.cpp (hunk 3 above)
Comment (Goober5000) [reply fred2/management.cpp:2090]: Well, I didn't want to make the diff any more complicated...

File: code/ai/aigoals.cpp (hunk 8 above)
Comment (Goober5000) [reply code/ai/aigoals.cpp:2550]: ok

**Review by BMagnu — APPROVED** [review]

**Pull request merged**